homeassistant/components/octoprint.py

```diff
@@ -11,43 +11,121 @@
+    vol.Optional(CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
 
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, has_all_unique_names, [vol.Schema({
         vol.Required(CONF_API_KEY): cv.string,
         vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_SSL, default=False): cv.boolean,
+        vol.Optional(CONF_PORT, default=80): cv.port,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
         vol.Optional(CONF_NUMBER_OF_TOOLS, default=0): cv.positive_int,
-        vol.Optional(CONF_BED, default=False): cv.boolean
-    }),
+        vol.Optional(CONF_BED, default=False): cv.boolean,
+        vol.Optional(CONF_SENSORS, default={}): SENSOR_SCHEMA,
+        vol.Optional(CONF_BINARY_SENSORS, default={}): BINARY_SENSOR_SCHEMA
+    })]),
 }, extra=vol.ALLOW_EXTRA)
 
 
 def setup(hass, config):
     """Set up the OctoPrint component."""
-    base_url = 'http://{}/api/'.format(config[DOMAIN][CONF_HOST])
-    api_key = config[DOMAIN][CONF_API_KEY]
-    number_of_tools = config[DOMAIN][CONF_NUMBER_OF_TOOLS]
-    bed = config[DOMAIN][CONF_BED]
-
-    hass.data[DOMAIN] = {"api": None}
-
-    try:
-        octoprint_api = OctoPrintAPI(base_url, api_key, bed, number_of_tools)
-        hass.data[DOMAIN]["api"] = octoprint_api
-        octoprint_api.get('printer')
-        octoprint_api.get('job')
-    except requests.exceptions.RequestException as conn_err:
-        _LOGGER.error("Error setting up OctoPrint API: %r", conn_err)
-
-    return True
+    printers = hass.data[DOMAIN] = {}
+    success = False
+
+    def device_discovered(service, info):
+        """Get called when an Octoprint server has been discovered."""
+        _LOGGER.debug('Found an Octoprint server: %s', info)
+
+    discovery.listen(hass, SERVICE_OCTOPRINT, device_discovered)
+
+    for printer in config[DOMAIN]:
+        name = printer[CONF_NAME]
+        ssl = 's' if printer[CONF_SSL] else ''
+        base_url = 'http{}://{}:{}/api/'.format(ssl,
+                                                printer[CONF_HOST],
+                                                printer[CONF_PORT])
+        api_key = printer[CONF_API_KEY]
+        number_of_tools = printer[CONF_NUMBER_OF_TOOLS]
+        bed = printer[CONF_BED]
+        try:
+            octoprint_api = OctoPrintAPI(base_url, api_key, bed,
+                                         number_of_tools)
+            printers[base_url] = octoprint_api
+            octoprint_api.get('printer')
+            octoprint_api.get('job')
+        except requests.exceptions.RequestException as conn_err:
+            _LOGGER.error("Error setting up OctoPrint API: %r", conn_err)
+            continue
+
+        sensors = printer[CONF_SENSORS][CONF_MONITORED_CONDITIONS]
+        load_platform(hass, 'sensor', DOMAIN, {'name': name,
```

Comment:
An optimization would be to merge the discovery_info for each printer into one container and call load_platform only once per domain. But it's not strictly necessary.